File: old/src/edu/oswego/csc436/states/Emergency.java
package edu.oswego.csc436.states;

import edu.oswego.csc436.Constants;
import edu.oswego.csc436.data.BadSensorValueException;
import edu.oswego.csc436.data.Data;

import java.io.IOException;

public class Emergency extends State {

  private static Emergency instance = new Emergency();

  @Override
  public State update(float dt, Data data) {
    try {
      if (data.getTargetSpeed() <= Float.MIN_NORMAL) return Alert.getInstance();
      data = data.update(dt, Math.max(0.0f, data.getTargetSpeed() - (dt * Constants.EMERGENCY_DECEL_RATE)));
      data.writeSpeeds();

      return this;
    } catch (BadSensorValueException | IOException e) {
      e.printStackTrace(System.err);
      return this;
    }
  }

  public static State getInstance() {
    return instance;
  }
}

File: src/state/emergency.py
import sys

from constants import Constants
from data import BadSensorValueError
from state import State
import alert
import follow
import gopigo.gopigo

class Emergency(State):
    _instance = None

    def __init__(self):
        pass

    def update(self, dt, data):
        try:
            data.set_speed(0.0)
            target_speed = data.get_target_speed()
            data.write_speed()
            gopigo.stop()
            if target_speed < 0.1:
                return alert.Alert.get_instance()

            return self

        except (IOError, BadSensorValueError) as e:
            sys.stderr.write(e)
            return self

    @staticmethod
    def get_instance():
        if Emergency._instance == None:
            Emergency._instance = Emergency()
        return Emergency._instance
